### PR TITLE
Add support for async requests for wasm plugin

### PR DESCRIPTION
**Pull request by shukitchan**

### CODE REVIEW

**Comment by shukitchan** [comment]
[approve ci autest]

**Comment by shukitchan** [comment]
[approve ci autest]

**Comment by shukitchan** [comment]
[approve ci autest]

**Comment by shukitchan** [comment]
There are quite a lot of materials on async requests in a wasm module in this talk - https://events.istio.io/istiocon-2021/slides/c8p-ExtendingEnvoyWasm-EdSnible.pdf

File: plugins/experimental/wasm/wasm_main.cc [hunk 1]
```diff
@@ -243,15 +246,25 @@ http_event_handler(TSCont contp, TSEvent event, void *data)
...
-  if (result == 0) {
-    TSHttpTxnReenable(txnp, TS_EVENT_HTTP_CONTINUE);
-  } else if (result < 0) {
-    TSHttpTxnReenable(txnp, TS_EVENT_HTTP_ERROR);
+  // check if we have reenable transaction already or not
+  if ((context == nullptr) || (!context->isTxnReenable())) {
+    TSDebug(WASM_DEBUG_TAG, "[%s] no context or not yet reenabled transaction", __FUNCTION__);
+
+    if (result == 0) {
+      TSHttpTxnReenable(txnp, TS_EVENT_HTTP_CONTINUE);
```
Comment (serrislew) [inline plugins/experimental/wasm/wasm_main.cc:254]: Why does `context->reenable_txn_` not need to get set here?

File: doc/admin-guide/plugins/wasm.en.rst [hunk 2]
```diff
@@ -114,5 +113,6 @@ The plugin will not support the following functionality as specified in Proxy-Wa
 
 * Getting and setting trailer request and response header
 * Getting and setting HTTP/2 frame meta data
+* Support asynchronous request call in the start handler function of the transaction lifecycle
```
Comment (serrislew) [inline doc/admin-guide/plugins/wasm.en.rst:116]: Should this be in limitations? I thought this is a new feature but adding it here implies it is not supported

Comment (shukitchan) [reply doc/admin-guide/plugins/wasm.en.rst:116]: Yes. This is a new feature but there is no easy way for us to support async request inside the "start" handler function of the plugin lifecycle. 

Let me reword this a bit.

File: plugins/experimental/wasm/wasm_main.cc (hunk 1 above)
Comment (shukitchan) [reply plugins/experimental/wasm/wasm_main.cc:254]: Yes. I should do that. Will make a change.

**Comment by bneradt** [comment]
[approve ci clang-format]

**Comment by shukitchan** [comment]
[approve ci autest]

**Comment by shukitchan** [comment]
[approve ci autest]

**Comment by shukitchan** [comment]
[approve ci autest]

File: plugins/experimental/wasm/ats_context.cc [hunk 3]
```diff
@@ -1533,11 +1708,18 @@ Context::getHeaderMap(WasmHeaderMapType type)
...
   case WasmHeaderMapType::ResponseTrailers:
     return {};
+  case WasmHeaderMapType::HttpCallResponseHeaders:
+    if (cr_hdr_buf_ == nullptr || cr_hdr_loc_ == nullptr) {
+      return {};
+    }
+    map.bufp    = cr_hdr_buf_;
+    map.hdr_loc = cr_hdr_loc_;
+    return map;
+  case WasmHeaderMapType::HttpCallResponseTrailers:
```
Comment (serrislew) [inline plugins/experimental/wasm/ats_context.cc:1718]: nit-picky: didn't need to be added. no change in functionality

**Review by serrislew — DISMISSED** [review]

**Review by serrislew — APPROVED** [review]

**Pull request merged**